### ISSUE TITLE
SSL connection error: CertCreateCertificateChainEngine failed. Invalid parameter. Error 2147942487/0x80070057

### ISSUE THREAD

**Issue by storm49152** [HeidiSQL/HeidiSQL#1766]
**Preconditions**
* Server: Rocky Linux 8.7
* MariaDB: 10.8.7
* OpenSSL: 1.1.1k

MariaDB configuration:

```ini
[mariadb] 
ssl = on 
ssl_cert = /path/to/mariadb-server.crt 
ssl_key = /path/to/mariadb-server.key 
ssl_ca = /path/to/ca-bundle.crt
```

* Client: Fedora 37
* MariaDB: 10.5.18
* OpenSSL: 3.0.8
* Wine: 8.1.1
* HeidiSQL: 12.3.0.6657

MariaDB configuration:

```ini
[mysql]
host = server.domain.tld
port = 33306
user = root
password = <super-secret-password>
ssl = on
ssl-ca = /path/to/ca-bundle.crt
ssl-cert = /path/to/mysql-client.crt
ssl-key = /path/to/mysql-client.key
max_allowed-packet = 1G
```

I know that I'm using a non-standard port: it is **_not_** a typo, and the server **_is_** available on that port.

I'm using a local root and intermediate CA. From the intermediate CA, I created both the server and client certificates. When I verify the server and client certificates against the CA bundle, OpenSSL says they're okay.

**Describe the bug**

I can connect using the `mysql` command:

```shell
[user@client-host ~]$ mysql -Nse "SHOW SESSION STATUS LIKE 'Ssl_accepts';"
Ssl_accepts     8
[user@client-host ~]$ mysql -Nse "SHOW SESSION STATUS LIKE 'Ssl_cipher';"
Ssl_cipher      TLS_AES_256_GCM_SHA384
[user@client-host ~]$ mysql -Nse "SHOW SESSION STATUS LIKE 'Ssl_version';"
Ssl_version     TLSv1.3
```

To me, this proves that the CA and certificates are fine, and that the SSL configuration works.

**To Reproduce**

When I configure HeidiSQL with the same CA bundle and client key/certificate, I'm consistently getting:

```text
SSL connection error: CertCreateCertificateChainEngine failed. Invalid parameter. Error 2147942487/0x80070057
```

I attached some screenshots from the HeidiSQL configuration and error.

If it helps, I can also provide Wireshark captures from succeeding `mysql` and failing HeidiSQL connections. The failing HeidiSQL connection is showing `Ignored Unknown Record` twice, a `TCP Out-Of-Order`, and a `TCP Retransmission` in the same data stream, none of which occur when using `mysql`.  (I'd like to keep those private). 

![HeidiSQL-connection](https://user-images.githubusercontent.com/6792545/219953007-1cb3972b-d631-4b84-bae7-0b7bf6ea1406.png)
![HeidiSQL-SSL](https://user-images.githubusercontent.com/6792545/219953011-189eb9e6-7fa4-4272-8e27-304167d04e69.png)
![HeidiSQL-error](https://user-images.githubusercontent.com/6792545/219953016-da538538-45a5-477e-842d-8434416264aa.png)

**Comment by ansgarbecker** [comment]
Please try the other libraries as well. The issue may be caused by libmariadb.dll. And of course retry with the latest v12.5.

**Comment by ansgarbecker** [comment]
Here's a comment saying the error is only present in the 32bit version of HeidiSQL, but not in the 64bit builds: https://aur.archlinux.org/packages/heidisql#comment-889051 . Can you confirm that?

**Comment by storm49152** [comment]
I'm using the 64bit version of HeidiSQL, and I'm currently at:

![image](https://github.com/HeidiSQL/HeidiSQL/assets/6792545/be4c32c2-48ec-4c83-8c14-18d63f6a0cdf)

Since I created this issue, I have a new server install and MariaDB SSL is no longer configured. However, I will do this again to test the other libraries as well (I think I did that, but it's too long ago to be sure), and report back.

**Comment by storm49152** [comment]
Server: MariaDB 11.0.2
HeidiSQL: 12.5.0.6685

I configured both MariaDB client and server for TLS, and AFAICT the configuration should be fine:

```
[user@vps ~]$ mariadb
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 49
Server version: 11.0.2-MariaDB MariaDB Server

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> SHOW GLOBAL VARIABLES LIKE 'have_ssl';
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| have_ssl      | YES   |
+---------------+-------+
1 row in set (0.001 sec)

MariaDB [(none)]> SHOW SESSION STATUS LIKE 'Ssl_cipher';
+---------------+------------------------+
| Variable_name | Value                  |
+---------------+------------------------+
| Ssl_cipher    | TLS_AES_256_GCM_SHA384 |
+---------------+------------------------+
1 row in set (0.001 sec)
MariaDB [(none)]> SHOW SESSION STATUS LIKE 'Ssl_version';
+---------------+---------+
| Variable_name | Value   |
+---------------+---------+
| Ssl_version   | TLSv1.3 |
+---------------+---------+
1 row in set (0.001 sec)
```

HeidiSQL SSL configuration:

![image](https://github.com/HeidiSQL/HeidiSQL/assets/6792545/c3d84f74-aa29-4cce-88ec-ac3d2cb3f80d)

The key, certificate and CA bundle used here are identical to the one used for the command line test.

The error message changed from my initial try. There is now the complaint that no valid certificates can be found in the CA bundle, even though this works on Linux. It doesn't matter if I change the file from Unix to DOS format. The `libmysql` libraries produce nothing useful at all.

Connecting using `libmariadb.dll`:

![image](https://github.com/HeidiSQL/HeidiSQL/assets/6792545/c95f7ebf-2198-4f61-b653-053bd4569d4a)

Connecting using `libmysql-6.1.dll`:

![image](https://github.com/HeidiSQL/HeidiSQL/assets/6792545/840ad728-38b7-4b8d-a0a4-f2ddcee9150e)

Connecting using `libmysql.dll`:

![image](https://github.com/HeidiSQL/HeidiSQL/assets/6792545/ffcb14c9-f902-4f46-9e96-185ac520179c)

**Comment by storm49152** [comment]
I searched for the error message (not specifically the error code, because that returns too many unrelated results), and that came up with this: [https://stackoverflow.com/questions/72794510/django-not-connecting-to-planetscale-ssl-error](https://stackoverflow.com/questions/72794510/django-not-connecting-to-planetscale-ssl-error). I'm unsure if the solution found there is helpful.

**Comment by ansgarbecker** [comment]
Thanks for your tests and research.
According to the documentation of [mysql_ssl_set](https://mariadb.com/kb/en/mysql_ssl_set/) (which HeidiSQL uses), the CA file is passed correctly to the library. One cause could be that the path is invalid under Wine, but that's just a guess.
The above mentioned post on StackOverflow points me to a different approach for preparing the MySQL connection: The C-API method [mysql_optionsv](https://mariadb.com/kb/en/mysql_optionsv/#tls-options) provides the same SSL options as `mysql_ssl_set` does, for example `MYSQL_OPT_SSL_CA`. HeidiSQL only uses the `mysql_ssl_set` method only, so I wonder if I have to use the options as well. But this is very unlikely, as other users are successfully using SSL with HeidiSQL.
Is your CA file _a PEM file that should contain one or more X509 certificates for trusted Certificate Authorities (CAs) to use for TLS_ (cited from the docs) ?
You could try to place a newer libmariadb.dll in your HeidiSQL. You can name it e.g. "libmariadb-11.dll", to have a version indicator. HeidiSQL should list it in the library dropdown.

**Comment by storm49152** [comment]
At first I chose a path on the Linux filesystem which could be invalid under Wine. However, Wine maps `/` to `Z:`, so all paths relative to `Z:\` (which I used) should be valid. Obviously; if the user has read access to the directory/file. However, I copied all key and cert files into the `drive_c/certs` folder, which inside of Wine is perceived as `C:\certs`, to make sure I was not running into that issue.

The CA file is the system CA bundle, and OpenSSL seems to think it is in PEM format:

```
$ openssl x509 -text -noout -in /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 2 (0x2)
        Signature Algorithm: sha256WithRSAEncryption
[...etc...]
```

The command only shows the first certificate in the file, but this is a bundle file that otherwise works. The client certificate shows similar output, so should also be valid PEM.
If the bundle actually is invalid, then the (Fedora) system command `update-ca-trust` command is failing, because that is what generates the file. And if that would be the case, then I would expect a lot of other things also fail, but I don't see that. FTR, this CA file and the client key/certificate combination works just fine with the command line `mysql`/`mariadb` utility.
The only problem I suspected was the file format: Unix line endings `\n` vs Windows line endings `\r\n`. So I tried again by changing the CA bundle, the client key and PEM CRT files to DOS format, but that didn't solve it.

I just tested using libmariadb.dll from MariaDB 11.0.2, and it fails with the same error message: "no valid certificates found".

**Comment by ansgarbecker** [comment]
This may be related to what I found recently, and which also caused issue #1426 . If you update HeidiSQL with the latest build installer from the nightly builds page, I highly guess you can connect then.
Note you need to use the ["32+64bit installer" link from Revision 12.7.0.6868](https://www.heidisql.com/download.php#nightlybuilds).
After updating, change the library in your session to "**libmysql-8.4.0.dll**" and try to connect.

**Comment by storm49152** [comment]
I finally got around to installing HeidiSQL 12.8.0.6908, no updates available. Next, I changed the library to "libmysql-8.4.0.dll". Tried to connect, and this time I get a popup that the certificate cannot be verified, which is a bummer, but more helpful.

![image](https://github.com/user-attachments/assets/78d62754-6bb3-4cb9-a7d8-0607c50fa28d)

I used `openssl verify -CAfile ... -untrusted ... cert.pem` to check the server and client certificates validity, and it says that both certificates are valid. Also, the end dates of all the root, intermediate, server and client certs are okay. But... that is what Linux says, not Wine.

The error message mentions to lower the certificate verification in SSL settings, so I've set it to "Verify CA (insecure)". This makes the error go away, and HeidiSQL connects to the server. It's the "Verify CA and host name identity (May fail with self-signed certificate)" setting that indeed fails.

I did import my CA root and intermediate certificates into the system CA store, and native applications are fine. But this is Wine, so, apparently, that doesn't help. I'm using HeidiSQL in a 64-bit Wine prefix, and I have yet to find a functioning (!) way to find a way to import the root and intermediate certs. It would have been nice to have "mmc.exe", but alas.

But that is not your problem, so I'm closing this ticket. Thanks for fixing this.
